sloghandler.go

```diff
@@ -0,0 +1,71 @@
+//go:build go1.21
+// +build go1.21
+
+/*
+Copyright 2023 The logr Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logr
+
+import (
+	"context"
+	"log/slog"
+)
+
+type slogHandler struct {
+	sink        LogSink
+	groupPrefix string
+}
+
+func (l *slogHandler) Enabled(ctx context.Context, level slog.Level) bool {
+	return l.sink.Enabled(levelFromSlog(level))
+}
+
+func (l *slogHandler) Handle(ctx context.Context, record slog.Record) error {
+	kvList := make([]any, 0, 2*record.NumAttrs())
+	record.Attrs(func(attr slog.Attr) bool {
+		kvList = append(kvList, appendPrefix(l.groupPrefix, attr.Key), attr.Value.Any())
+		return true
+	})
+	if record.Level >= slog.LevelError {
+		l.sink.Error(nil, record.Message, kvList...)
+	} else {
+		l.sink.Info(levelFromSlog(record.Level), record.Message, kvList...)
+	}
+	return nil
+}
+
+func (l slogHandler) WithAttrs(attrs []slog.Attr) slog.Handler {
+	kvList := make([]any, 0, 2*len(attrs))
+	for _, attr := range attrs {
+		kvList = append(kvList, appendPrefix(l.groupPrefix, attr.Key), attr.Value.Any())
+	}
+	l.sink = l.sink.WithValues(kvList...)
+	return &l
+}
+
+func (l slogHandler) WithGroup(name string) slog.Handler {
```

Comment:
> I thought it would at least assemble a struct and log it as such

That's indeed what the slog.JSONLogger does:
```
package main

import (
	"log/slog"
	"os"
)

func main() {
	logger := slog.New(slog.NewJSONHandler(os.Stdout, nil))
	logger.WithGroup("group").With("hello", "world").Info("ping")
}
```

=>

```
{"time":"2023-08-04T08:33:43.56963767+02:00","level":"INFO","msg":"ping","group":{"hello":"world"}}
```

When wrapping zapr (the current one without slog support), I get instead:
```
{"level":"info","msg":"ping","group.hello":"world"}
```

With [slog support](https://github.com/go-logr/zapr/pull/60) there is at least a chance to implement `WithGroup` properly - but I haven't tried that yet (open TODO).

Reply:
Once I found https://github.com/uber-go/zap/blob/v1.25.0/field.go#L334 it was trivial.

Now I get:
```
{"level":"info","msg":"ping","group":{"hello":"world"}}
```

Reply:
How does that intersect with WithValues() ?

E.g.:

```
s := logr.ToSlog(myLogger)
s.With("a", 1)  // calls Handler.WithAttrs(), which calls logr WithValues("a", 1)
s = s.WithGroup("g")
s.With("b", 2)  // calls Handler.WithAttrs(), which calls logr WithValues("g.b", 2)
s.Info("msg", "c", 3) // calls Handler.Handle, which calls logr.Info("msg", "g.c", 3)
```

The only options I see are

0) Live with output `"g.b": 2, "g.c": 3`

1) don't use logrHandler under slog, use native slog support

2) accumulate calls to WithAttr(), so it becomes:
```
s := logr.ToSlog(myLogger)
s.With("a", 1)  // calls Handler.WithAttrs(), which calls logr WithValues("a", 1)
s = s.WithGroup("g")
s.With("b", 2)  // calls Handler.WithAttrs(), which saves "b"=2
s.Info("msg", "c", 3) // calls Handler.Handle, which calls logr.Info("msg", "g", map[string]any{"b": 2, "c": 3})
```

3) teach logr to have a "group" construct.

I'm interested in this approach, but I think we can accept (0) until we do that.  Fair?

https://github.com/go-logr/logr/issues/199

Reply:
Yes.